### PR TITLE
Major refactoring

### PR DESCRIPTION
**Pull request by qzhuyan**
# Why we need major refactoring?

## Sync to most recent MsQuic version
Most of the current Quicer code base was based on MsQuic version 0.9. 
There were some updates for adapting to the newer MsQuic version(1.x and 2.x) but most are just modified to compile and pass the existing tests. 

## Move from a TCP/TLS replacement to a Feature Rich Quic Stack.  
In newer MsQuic, especially in documents, there are more explanations about the API, flags, callbacks that we may/may not deal with them correctly in QUIC protocol layer since the PoC version mostly just try to replace TCP/TLS without utilize all the features/rich information from QUIC protocol.

There was some ad-hoc code changes to deal with the signals from MsQuic Stack, now we just want to make it more general. 

## Go production grade
With current impl. we have verified the performance of the current implementation (handshake,  latency, throughput, and resource usage )  Also did some memory leak checks with existing testcases. Also we have EMQX users that proves Quic 
Protocol has improved connectiviy in their live networks. 

We could assume we have a good base so far and could start adding more stuff to make it production-grade.

## Provides some general behavior to make it easy to develop with quicer
It is not easy to manage the relationship of QUIC Connecion and Streams in terms of life cycle, especially when comes to the multi-stream support.
Also with current existing APIs, one cannot understand how to use them without looking into the QUIC protocol.

We want to have a better document also may provide some general behavior to make it easier.

### CODE REVIEW

**Comment by essen** [comment]
Hello, do you have a rough timeline for merging this? I need unidirectional streams for HTTP/3. Thanks for the hard work!

**Comment by qzhuyan** [comment]
> Hello, do you have a rough timeline for merging this? I need unidirectional streams for HTTP/3. Thanks for the hard work!

by the end of this week.

File: src/quicer_stream.erl [hunk 1]
```diff
@@ -124,68 +212,132 @@ handle_cast(_Request, State) ->
...
+          {noreply, state()} |
+          {noreply, state(), Timeout :: timeout()} |
+          {noreply, state(), hibernate} |
+          {stop, Reason :: normal | term(), state()}.
+handle_info({quic, new_stream, Stream, Flags},
+            #{ stream_opts := Options
+             , stream := undefined
+             , callback := CallbackModule
+             , callback_state := CBState
+             } = State) ->
```
Comment (qzhuyan) [inline src/quicer_stream.erl:225]: comment, for stream acceptor.

File: src/quicer_stream.erl [hunk 2]
```diff
@@ -61,44 +122,71 @@ start_link(Conn, Opts) ->
...
+%% After conn handshake, with stream handle
+init([Stream, Conn, StreamOpts]) when is_list(StreamOpts) ->
+    ?tp(new_stream_2, #{module=>?MODULE, stream=>Stream}),
+    init([Stream, Conn, maps:from_list(StreamOpts)]);
+init([Stream, Conn, #{stream_callback := CallbackModule} = StreamOpts]) ->
+    ?tp(new_stream_3, #{module=>?MODULE, stream=>Stream}),
+    process_flag(trap_exit, true),
+    case CallbackModule:new_stream(Stream, StreamOpts, undefined) of
+        {ok, CallbackState} ->
+            State = #{ is_owner => false
```
Comment (qzhuyan) [inline src/quicer_stream.erl:154]: move to fun: `init_state`

File: src/quicer_stream.erl [hunk 3]
```diff
@@ -124,68 +212,132 @@ handle_cast(_Request, State) ->
...
+             , callback := CallbackModule
+             , callback_state := CBState
+             } = State) ->
+    ?tp(new_stream, #{module=>?MODULE, stream=>Stream, stream_flags => Flags}),
+    try CallbackModule:new_stream(Stream, Options#{open_flags => Flags}, CBState) of
+        {ok, CallbackState} ->
+            {noreply, State#{stream => Stream, callback_state => CallbackState}};
         {error, Reason} ->
-            {stop, Reason, State#state{stream = Stream}}
+            {stop, Reason, State#{stream => Stream}}
```
Comment (qzhuyan) [inline src/quicer_stream.erl:231]: :=

File: src/quicer_stream.erl [hunk 4]
```diff
@@ -124,68 +212,132 @@ handle_cast(_Request, State) ->
...
+handle_info({quic, new_stream, Stream, Flags},
+            #{ stream_opts := Options
+             , stream := undefined
+             , callback := CallbackModule
+             , callback_state := CBState
+             } = State) ->
+    ?tp(new_stream, #{module=>?MODULE, stream=>Stream, stream_flags => Flags}),
+    try CallbackModule:new_stream(Stream, Options#{open_flags => Flags}, CBState) of
+        {ok, CallbackState} ->
+            {noreply, State#{stream => Stream, callback_state => CallbackState}};
```
Comment (qzhuyan) [inline src/quicer_stream.erl:229]: :=

File: src/quicer_stream.erl [hunk 5]
```diff
@@ -15,24 +15,76 @@
...
-               , cbstate :: any()
-               }).
-
--type stream_opts() :: map().
+-type state() :: #{ stream := quicer:stream_handle()
+                  , conn := quicer:connection_handle()
+                  , callback := atom()
+                  , callback_state := term()
+                  , is_owner := boolean()
+                  , stream_opts := map()
```
Comment (qzhuyan) [inline src/quicer_stream.erl:86]: add open_flag & start_flag

File: src/quicer_stream.erl [hunk 6]
```diff
@@ -124,68 +212,132 @@ handle_cast(_Request, State) ->
...
         _:Reason:ST ->
             maybe_log_stracetrace(ST),
             {stop, {handle_stream_data_crash, Reason}, State}
     end;
 
-handle_info({quic, _Bin, StreamA, _, _, _}, #state{stream = StreamB} = State)
-  when StreamB =/=StreamA ->
-    ?tp(inval_stream_data, #{module=>?MODULE, stream_a=>StreamA, stream_b => StreamB}),
-    {stop, wrong_stream, State};
+handle_info({quic, start_completed, Stream,
```
Comment (qzhuyan) [inline src/quicer_stream.erl:255]: consider rename `start_completed` to `start_complete`

File: src/quicer_stream.erl [hunk 7]
```diff
@@ -249,3 +401,10 @@ is_fin(0) ->
     false;
 is_fin(Flags) when is_integer(Flags) ->
     (1 bsl 1) band Flags =/= 0.
+
+%% handoff must happen
+wait_for_handoff() ->
+    receive
```
Comment (qzhuyan) [inline src/quicer_stream.erl:407]: monitor then wait, also handle EXIT

File: src/quicer_stream.erl [hunk 8]
```diff
@@ -124,68 +212,132 @@ handle_cast(_Request, State) ->
...
+%% This callback is optional, so callback modules need to export it only if they return
+%% one of the tuples containing {continue,Continue} from another callback.
+%% If such a {continue,_} tuple is used and the callback is not implemented,
+%% the process will exit with undef error.
+-spec handle_continue(Continue::term(), State::term()) ->
+          {noreply, state()} |
+          {noreply, state(), Timeout :: timeout()} |
+          {noreply, state(), hibernate} |
+          {stop, Reason :: normal | term(), state()}.
+handle_continue(?post_init, #{ is_owner := false, stream := Stream} = State) ->
```
Comment (qzhuyan) [inline src/quicer_stream.erl:325]: change `post_init` to `wait_for_handoff`

File: src/quicer_stream.erl [hunk 9]
```diff
@@ -124,68 +212,132 @@ handle_cast(_Request, State) ->
...
+             , callback_state := CBState} = State) ->
+    ?tp(debug, #{module=>?MODULE, event => peer_accepted}),
+    default_cb_ret(M:peer_accepted(Stream, CBState), State);
+
+handle_info({quic, passive, Stream, undefined},
+            #{ callback := M
+             , callback_state := CBState} = State) ->
+    ?tp(debug, #{module=>?MODULE, event => passive}),
+    default_cb_ret(M:passive(Stream, CBState), State).
+
```
Comment (qzhuyan) [inline src/quicer_stream.erl:312]: handle EXIT from the connection owner and then make a callback

File: src/quicer_connection.erl [hunk 10]
```diff
@@ -0,0 +1,386 @@
...
+             , callback_state := CbState} = State) ->
+    ?tp(debug, #{module => ?MODULE, conn => C, event => peer_address_changed, new_addr => NewAddr}),
+    default_cb_ret(M:peer_address_changed(C, NewAddr, CbState), State);
+
+handle_info({quic, new_stream, Stream, Flags},
+            #{ conn := C
+             , callback := M
+             , callback_state := CbState} = State) when C =/= undefined->
+    %% Best practice:
+    %%   One connection will have a control stream that have the same life cycle as the connection.
```
Comment (qzhuyan) [inline src/quicer_connection.erl:279]: s/have/has

File: src/quicer_connection.erl [hunk 11]
```diff
@@ -0,0 +1,386 @@
...
+    %%   The connection may spawn one *control stream* acceptor before starting the handshake
+    %%   AND the stream acceptor should accept new stream so it will likely pick up the control stream
+    %% note, by desgin, control stream doesn't have to be the first stream initiated.
+    %% here, it handles new stream when there is no available stream acceptor for the connection.
```
Comment (qzhuyan) [inline src/quicer_connection.erl:283]: new streams

File: c_src/quicer_listener.c [hunk 12]
```diff
@@ -82,19 +82,81 @@ ServerListenerCallback(__unused_parm__ HQUIC Listener,
...
+        ATOM_SERVER_NAME,   // server name
+        ATOM_ALPNS,         // alpns
+        ATOM_CLIENT_ALPNS,  // client alpns
+        ATOM_CRYPTO_BUFFER, // crypto buffer
+      };
+
+      ERL_NIF_TERM props_value[] = {
+        enif_make_uint(env,
+                       Event->NEW_CONNECTION.Info->QuicVersion), // version
+        enif_make_string(
```
[the remark this answers is not in the document]
Comment (qzhuyan) [reply c_src/quicer_listener.c:136]: TBD

File: c_src/quicer_nif.c [hunk 13]
```diff
@@ -1158,6 +1248,51 @@ stream_controlling_process(ErlNifEnv *env,
...
+                      size_t cnt)
+{
+  ERL_NIF_TERM prop = ATOM_UNDEFINED;
+
+  // build prop
+  if (!(0 == cnt || values == NULL || keys == NULL))
+    {
+      assert(values != NULL);
+      assert(keys != NULL);
+      enif_make_map_from_arrays(env, keys, values, cnt, &prop);
```
[the remark this answers is not in the document]
Comment (qzhuyan) [reply c_src/quicer_nif.c:1270]: prop will become `undefined` and we will know.

File: src/quicer_server_conn_callback.erl [hunk 14]
```diff
@@ -15,28 +15,172 @@
...
-    ok = quicer:async_handshake(Conn),
+closed(_Conn, #{} = _Flags, S)->
     {ok, S}.
 
-connected(Conn, #{slow_start := false, stream_opts := SOpts} = S) ->
-    quicer_stream:start_link(Conn, SOpts),
-    {ok, S};
-connected(_Conn, S) ->
+new_conn(Conn, #{version := _Vsn}, #{stream_opts := SOpts} = S) ->
+    %% @TODO configurable behavior of spawning stream acceptor
```
[the remark this answers is not in the document]
Comment (qzhuyan) [reply src/quicer_server_conn_callback.erl:64]: added in todo

File: c_src/quicer_connection.c [hunk 15]
```diff
@@ -1058,6 +940,395 @@ handle_dgram_recv_event(QuicerConnCTX *c_ctx, QUIC_CONNECTION_EVENT *Event)
...
+        }
+      // We must copy here, otherwise it will become double free
+      // in resource dealloc callbacks (for Stream and Connection)
+      memcpy(acc, c_ctx->owner, sizeof(ACCEPTOR));
+
+      // @TODO we could set it to passive and let new owner set it to
+      // active
+      //  then old owner and new owner do not need to do handoff
+      //  but that will buffer more in quic stack and hit limit control
+      //  credit unsure what is the better solution.
```
[the remark this answers is not in the document]
Comment (qzhuyan) [reply c_src/quicer_connection.c:1134]: become todo

File: c_src/quicer_connection.c [hunk 16]
```diff
@@ -1058,6 +940,395 @@ handle_dgram_recv_event(QuicerConnCTX *c_ctx, QUIC_CONNECTION_EVENT *Event)
...
+static QUIC_STATUS
+handle_connection_event_peer_needs_streams(
+    QuicerConnCTX *c_ctx, __unused_parm__ QUIC_CONNECTION_EVENT *Event)
+{
+  assert(QUIC_CONNECTION_EVENT_PEER_NEEDS_STREAMS == Event->Type);
+  assert(c_ctx->Connection);
+  ErlNifEnv *env = c_ctx->env;
+  ERL_NIF_TERM report = make_event(env,
+                                   ATOM_PEER_NEEDS_STREAMS,
+                                   enif_make_resource(env, c_ctx),
```
[the remark this answers is not in the document]
Comment (qzhuyan) [reply c_src/quicer_connection.c:1196]: become todo

File: c_src/quicer_connection.c [hunk 17]
```diff
@@ -1058,6 +940,395 @@ handle_dgram_recv_event(QuicerConnCTX *c_ctx, QUIC_CONNECTION_EVENT *Event)
...
+
+  ERL_NIF_TERM props_name[] = { ATOM_BIDI_STREAMS, ATOM_UNIDI_STREAMS };
+  ERL_NIF_TERM props_value[]
+      = { enif_make_uint64(env, Event->STREAMS_AVAILABLE.BidirectionalCount),
+          enif_make_uint64(env,
+                           Event->STREAMS_AVAILABLE.UnidirectionalCount) };
+
+  ERL_NIF_TERM report = make_event_with_props(env,
+                                              ATOM_STREAMS_AVAILABLE,
+                                              enif_make_resource(env, c_ctx),
```
[the remark this answers is not in the document]
Comment (qzhuyan) [reply c_src/quicer_connection.c:1179]: become todo

File: test/quicer_SUITE.erl [hunk 18]
```diff
@@ -1510,11 +1573,324 @@ tc_get_stream_rid(Config) ->
...
+  {ok, _QuicApp} = quicer:start_listener(mqtt, Port, Options),
+  {ok, Conn} = quicer:connect("localhost", Port, default_conn_opts(), 5000),
+  {ok, Stm} = quicer:start_stream(Conn, [ {active, 3}, {start_flag, ?QUIC_STREAM_START_FLAG_IMMEDIATE}
+                                        , {quic_event_mask, ?QUICER_STREAM_EVENT_MASK_SEND_COMPLETE}
+                                        ]),
+  {ok, Rid} = quicer:get_stream_rid(Stm),
+  %% We don't need to send anything, we should get start_completed even it is rate limited
+  receive
+    {quic, start_completed, Stm,
+             #{status := stream_limit_reached, stream_id := StreamID}} ->
```
[the remark this answers is not in the document]
Comment (qzhuyan) [reply test/quicer_SUITE.erl:1672]: become todo

File: src/quicer.erl [hunk 19]
```diff
@@ -203,7 +209,7 @@ connect(Host, Port, Opts, Timeout) when is_map(Opts) ->
   case quicer_nif:async_connect(Host, Port, NewOpts) of
     {ok, H} ->
       receive
-        {quic, connected, H} ->
+        {quic, connected, H, _} ->
```
[the remark this answers is not in the document]
Comment (qzhuyan) [reply src/quicer.erl:212]: todo

**Comment by qzhuyan** [comment]
left some todos and will be addressed in followup PRs. 
Also we need to check the coverage.

File: src/quicer_connection.erl [hunk 20]
```diff
@@ -0,0 +1,501 @@
...
+            {ok, State1#{callback_state => CBState}};
+        {ok, CBState, Action} ->
+            {ok, State1#{callback_state => CBState}, Action};
+         Other -> %% ignore, {stop, Reason} ...
+            Other
+    end;
+
+%% For Server
+init([undefined, Listener, {LOpts, COpts, SOpts}, Sup]) when is_list(COpts) ->
+    init([undefined, Listener, {LOpts, maps:from_list(COpts), SOpts}, Sup]);
```
Comment (id) [inline src/quicer_connection.erl:225]: Consider this
```
53> LOptions1 = [{conn_acceptors, 32} | LOptions].
[{conn_acceptors,32},
 {cert,"cert.pem"},
 {key,"key.pem"},
 {alpn,["sample"]},
 {peer_bidi_stream_count,1}]
54> quicer:start_listener(foobar, 1111, {LOptions1, [], []}).
=CRASH REPORT==== 13-Oct-2022::12:19:11.651324 ===
  crasher:
    initial call: quicer_connection:init/1
    pid: <0.787.0>
    registered_name: []
    exception error: undefined function undefined:init/1
      in function  quicer_connection:init/1 (/Users/id/work/quic/src/quicer_connection.erl, line 240)
      in call from gen_server:init_it/2 (gen_server.erl, line 423)
      in call from gen_server:init_it/6 (gen_server.erl, line 390)
    ancestors: [<0.786.0>,foobar,quicer_listener_sup,quicer_sup,<0.655.0>]
    message_queue_len: 0
    messages: []
    links: [<0.786.0>]
    dictionary: []
    trap_exit: true
    status: running
    heap_size: 610
    stack_size: 28
    reductions: 229
  neighbours:

{error,{{{badmatch,{error,{undef,[{undefined,init,
                                             [#{stream_opts => []}],
                                             []},
                                  {quicer_connection,init,1,
                                                     [{file,"/Users/id/work/quic/src/quicer_connection.erl"},
                                                      {line,240}]},
                                  {gen_server,init_it,2,[{file,"gen_server.erl"},{line,423}]},
                                  {gen_server,init_it,6,[{file,"gen_server.erl"},{line,390}]},
                                  {proc_lib,init_p_do_apply,3,
                                            [{file,"proc_lib.erl"},{line,226}]}]}}},
         [{quicer_listener,'-init/1-lc$^0/1-0-',2,
                           [{file,"/Users/id/work/quic/src/quicer_listener.erl"},
                            {line,94}]},
          {quicer_listener,init,1,
                           [{file,"/Users/id/work/quic/src/quicer_listener.erl"},
                            {line,94}]},
          {gen_server,init_it,2,[{file,"gen_server.erl"},{line,423}]},
          {gen_server,init_it,6,[{file,"gen_server.erl"},{line,390}]},
          {proc_lib,init_p_do_apply,3,
                    [{file,"proc_lib.erl"},{line,226}]}]},
        {child,undefined,
               {quicer_listener,foobar},
               {quicer_listener,start_link,
                                [foobar,1111,
                                 {[{conn_acceptors,32},
                                   {cert,"cert.pem"},
                                   {key,"key.pem"},
                                   {alpn,["sample"]},
                                   {peer_bidi_stream_count,1}],
                                  [],[]}]},
               transient,false,infinity,supervisor,
               [quicer_listener]}}}
```

File: test/quicer_SUITE.erl [hunk 21]
```diff
@@ -410,6 +427,35 @@ tc_conn_basic_slow_start(Config)->
...
+  {ok, {_, _}} = quicer:sockname(Conn),
+  ok = quicer:close_connection(Conn),
+  ok.
+
+tc_conn_basic_verify_peer_no_cacert(Config)->
+  %% Verify that the connection handshake should fail if
+  %% `verif_peer` is set but CA is unknown
+  Port = select_port(),
+  Owner = self(),
+  {SPid, Ref} = spawn_monitor(
```
[the remark this answers is not in the document]
Comment (qzhuyan) [reply test/quicer_SUITE.erl:452]: that reminds me forget to teardown the server, thanks.

File: test/quicer_SUITE.erl [hunk 22]
```diff
@@ -1318,7 +1381,7 @@ tc_setopt_conn_local_addr_in_use(Config) ->
   ct:pal("send after migration failed"),
   {ok, 5} = quicer:send(Stm0, <<"ping2">>),
```
[the remark this answers is not in the document]
Comment (qzhuyan) [reply test/quicer_SUITE.erl:1393]: The test is broken and it is disabled due to limitation in MsQuic, I added a mark.

File: src/quicer_connection.erl (hunk 20 above)
Comment (qzhuyan) [reply src/quicer_connection.erl]: I added a check in start_link

**Review by id — APPROVED** [review]
💯

**Comment by qzhuyan** [comment]
thanks @id you are brave enough to put your name on this giant PR.

Lets merge and continue improve it. ❤️

**Pull request merged**